File: 2022/day02-2.ts
function solution(content: string) {
  return content
    .split("\n")
    .map(parse)
    .map(play)
    .reduce((acc, cur) => acc + cur, 0);
}

function play([opponent, player]: string[]) {
  return (
    (player === "R" ? 1 : player === "P" ? 2 : 3) +
    (isLosing(opponent, player) ? 0 : opponent === player ? 3 : 6)
  );
}

function isLosing(opponent: string, player: string) {
  return (
    (opponent === "R" && player === "S") ||
    (opponent === "S" && player === "P") ||
    (opponent === "P" && player === "R")
  );
}

function isRock(opponent: string, player: string) {
  return (
    (opponent === "B" && player === "X") ||
    (opponent === "A" && player === "Y") ||
    (opponent === "C" && player === "Z")
  );
}

function isPaper(opponent: string, player: string) {
  return (
    (opponent === "C" && player === "X") ||
    (opponent === "B" && player === "Y") ||
    (opponent === "A" && player === "Z")
  );
}

function parse(line: string) {
  const [o, p] = line.split(" ");
  return [
    o === "A" ? "R" : o === "B" ? "P" : "S",
    isRock(o, p) ? "R" : isPaper(o, p) ? "P" : "S",
  ];
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2022/day02_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return sum(play(parse(line)) for line in content)


def play(hands: tuple[str, str]):
    opponent, player = hands
    return (0 if player == "X" else 3 if player == "Y" else 6) + (
        1
        if (opponent, player) in [("P", "X"), ("R", "Y"), ("S", "Z")]
        else 2 if (opponent, player) in [("S", "X"), ("P", "Y"), ("R", "Z")] else 3
    )


def parse(line: str):
    opponent, player = line.strip().split(" ")
    return ("R" if opponent == "A" else "P" if opponent == "B" else "S", player)


def main() -> None:
    with open("day02.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
